program/src/processor.rs

```diff
@@ -355,7 +356,113 @@ fn process_initialize_holder_rewards(
+        }
+
+        token_account.base.amount
+    };
+
+    let current_total_rewards = {
+        // Ensure the holder rewards pool is owned by the Paladin Rewards
+        // program.
+        if !holder_rewards_pool_info.owner.eq(program_id) {
+            return Err(ProgramError::InvalidAccountOwner);
+        }
+
+        // Ensure the provided holder rewards pool address is the correct
+        // address derived from the mint.
+        if !holder_rewards_pool_info
+            .key
+            .eq(&get_holder_rewards_pool_address(mint_info.key))
+        {
+            return Err(PaladinRewardsError::IncorrectHolderRewardsPoolAddress.into());
+        }
+
+        let holder_rewards_pool_data = holder_rewards_pool_info.try_borrow_data()?;
+        let holder_rewards_pool_state =
+            bytemuck::try_from_bytes::<HolderRewardsPool>(&holder_rewards_pool_data)
+                .map_err(|_| ProgramError::InvalidAccountData)?;
+
+        holder_rewards_pool_state.total_rewards
+    };
+
+    let (last_seen_total_rewards, unharvested_rewards) = {
+        // Ensure the holder rewards account is owned by the Paladin Rewards
+        // program.
+        if !holder_rewards_info.owner.eq(program_id) {
+            return Err(ProgramError::InvalidAccountOwner);
+        }
+
+        // Ensure the provided holder rewards address is the correct address
+        // derived from the token account.
+        if !holder_rewards_info
+            .key
+            .eq(&get_holder_rewards_address(token_account_info.key))
+        {
+            return Err(PaladinRewardsError::IncorrectHolderRewardsAddress.into());
+        }
+
+        let mut holder_rewards_data = holder_rewards_info.try_borrow_mut_data()?;
+        let holder_rewards_state =
+            bytemuck::try_from_bytes_mut::<HolderRewards>(&mut holder_rewards_data)
+                .map_err(|_| ProgramError::InvalidAccountData)?;
+
+        // Update the holder rewards state with the new "last seen" total
+        // rewards and zero out the unharvested rewards.
+        (
+            std::mem::replace(
+                &mut holder_rewards_state.last_seen_total_rewards,
+                current_total_rewards,
+            ),
+            std::mem::take(&mut holder_rewards_state.unharvested_rewards),
+        )
+    };
+
+    // Calculate unharvested rewards for the token account.
+    // Since the holder rewards account may already have unharvested rewards,
+    // calculate the share of rewards that have not been seen by the holder
+    // rewards account.
+    let unseen_total_rewards = current_total_rewards
+        .checked_sub(last_seen_total_rewards)
+        .ok_or(ProgramError::ArithmeticOverflow)?;
+    let unseen_unharvested_rewards =
+        calculate_reward_share(token_supply, token_account_balance, unseen_total_rewards)?;
+    let rewards_to_harvest = unharvested_rewards
+        .checked_add(unseen_unharvested_rewards)
+        .ok_or(ProgramError::ArithmeticOverflow)?;
```

Comment:
This doesn't need to be done now, but to be sure that the calculations here won't have any issues for all combinations of token_supply, token_account_balance, total_rewards, unharvested_rewards, and lamports in the actual account, we should add at least one proptest.

https://proptest-rs.github.io/proptest/ has some great info, and we have a few of these tests like https://github.com/solana-labs/solana-program-library/blob/61794be356d80a02023eeddce326a71727f7b8c4/token/program-2022/src/extension/transfer_fee/mod.rs#L428

Reply:
Gotcha. I've made a note to add regressions to all processor functions with calculations.

Reply:
Ok great! Typically, the most important checks are minimums (ie 0), maximums (u64::MAX, 100%, etc), and anywhere behavior might change, through truncating or flooring of numbers. The proptests are great at finding all of those limits. But they can also be very frustrating when there's a lot of math :sweat_smile: